File: database/src/resource/_schema.rs
use sqlx::types::chrono::{DateTime, Utc};
use uuid::Uuid;
use crate::common::type_value::{DataType, DataValue};

#[derive(Debug, Default, PartialEq, Clone)]
pub struct ModelSchema {
    pub id: Uuid,
    pub name: String,
    pub category: String,
    pub description: String,
    pub data_type: Vec<DataType>,
    pub tags: Vec<i16>,
    pub configs: Vec<Vec<ModelConfigSchema>>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct TagSchema {
    pub model_id: Uuid,
    pub tag: i16,
    pub name: String,
    pub members: Vec<i16>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct ModelConfigSchema {
    pub id: i32,
    pub model_id: Uuid,
    pub index: i16,
    pub name: String,
    pub value: DataValue,
    pub category: String
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct DeviceSchema {
    pub id: Uuid,
    pub gateway_id: Uuid,
    pub serial_number: String,
    pub name: String,
    pub description: String,
    pub type_id: Uuid,
    pub type_name: String,
    pub model_ids: Vec<Uuid>,
    pub configs: Vec<DeviceConfigSchema>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct GatewaySchema {
    pub id: Uuid,
    pub serial_number: String,
    pub name: String,
    pub description: String,
    pub type_id: Uuid,
    pub type_name: String,
    pub model_ids: Vec<Uuid>,
    pub configs: Vec<GatewayConfigSchema>
}

impl From<DeviceSchema> for GatewaySchema {
    fn from(value: DeviceSchema) -> Self {
        Self {
            id: value.gateway_id,
            serial_number: value.serial_number,
            name: value.name,
            description: value.description,
            type_id: value.type_id,
            type_name: value.type_name,
            model_ids: value.model_ids,
            configs: value.configs.into_iter().map(|el| el.into()).collect()
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct TypeSchema {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub model_ids: Vec<Uuid>,
    pub configs: Vec<TypeConfigSchema>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct DeviceConfigSchema {
    pub id: i32,
    pub device_id: Uuid,
    pub name: String,
    pub value: DataValue,
    pub category: String
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct GatewayConfigSchema {
    pub id: i32,
    pub gateway_id: Uuid,
    pub name: String,
    pub value: DataValue,
    pub category: String
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct TypeConfigSchema {
    pub id: i32,
    pub type_id: Uuid,
    pub name: String,
    pub value_type: DataType,
    pub value_default: DataValue,
    pub category: String
}

impl From<DeviceConfigSchema> for GatewayConfigSchema {
    fn from(value: DeviceConfigSchema) -> Self {
        Self {
            id: value.id,
            gateway_id: value.device_id,
            name: value.name,
            value: value.value,
            category: value.category
        }
    }
}

pub(crate) struct GroupSchema {
    pub(crate) id: Uuid,
    pub(crate) name: String,
    pub(crate) category: String,
    pub(crate) description: String,
    pub(crate) members: Vec<Uuid>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct GroupModelSchema {
    pub id: Uuid,
    pub name: String,
    pub category: String,
    pub description: String,
    pub model_ids: Vec<Uuid>
}

impl From<GroupSchema> for GroupModelSchema {
    fn from(value: GroupSchema) -> Self {
        Self {
            id: value.id,
            name: value.name,
            category: value.category,
            description: value.description,
            model_ids: value.members
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct GroupDeviceSchema {
    pub id: Uuid,
    pub name: String,
    pub category: String,
    pub description: String,
    pub device_ids: Vec<Uuid>
}

impl From<GroupSchema> for GroupDeviceSchema {
    fn from(value: GroupSchema) -> Self {
        Self {
            id: value.id,
            name: value.name,
            category: value.category,
            description: value.description,
            device_ids: value.members
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct GroupGatewaySchema {
    pub id: Uuid,
    pub name: String,
    pub category: String,
    pub description: String,
    pub gateway_ids: Vec<Uuid>
}

impl From<GroupSchema> for GroupGatewaySchema {
    fn from(value: GroupSchema) -> Self {
        Self {
            id: value.id,
            name: value.name,
            category: value.category,
            description: value.description,
            gateway_ids: value.members
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SetSchema {
    pub id: Uuid,
    pub template_id: Uuid,
    pub name: String,
    pub description: String,
    pub members: Vec<SetMember>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SetMember {
    pub device_id: Uuid,
    pub model_id: Uuid,
    pub data_index: Vec<u8>
}

pub(crate) struct SetMemberSort {
    pub(crate) device_id: Uuid,
    pub(crate) model_id: Uuid,
    pub(crate) data_index: Vec<u8>,
    pub(crate) set_position: i16
}

impl From<SetMemberSort> for SetMember {
    fn from(value: SetMemberSort) -> Self {
        Self {
            device_id: value.device_id,
            model_id: value.model_id,
            data_index: value.data_index
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SetTemplateSchema {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub members: Vec<SetTemplateMember>
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SetTemplateMember {
    pub type_id: Uuid,
    pub model_id: Uuid,
    pub data_index: Vec<u8>
}

pub(crate) struct SetTemplateMemberSort {
    pub(crate) type_id: Uuid,
    pub(crate) model_id: Uuid,
    pub(crate) data_index: Vec<u8>,
    pub(crate) template_index: i16
}

impl From<SetTemplateMemberSort> for SetTemplateMember {
    fn from(value: SetTemplateMemberSort) -> Self {
        Self {
            type_id: value.type_id,
            model_id: value.model_id,
            data_index: value.data_index
        }
    }
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct DataSchema {
    pub device_id: Uuid,
    pub model_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub data: Vec<DataValue>,
    pub tag: i16
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct DataSetSchema {
    pub set_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub data: Vec<DataValue>,
    pub tag: i16
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct BufferSchema {
    pub id: i32,
    pub device_id: Uuid,
    pub model_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub data: Vec<DataValue>,
    pub tag: i16
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct BufferSetSchema {
    pub ids: Vec<i32>,
    pub set_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub data: Vec<DataValue>,
    pub tag: i16
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SliceSchema {
    pub id: i32,
    pub device_id: Uuid,
    pub model_id: Uuid,
    pub timestamp_begin: DateTime<Utc>,
    pub timestamp_end: DateTime<Utc>,
    pub name: String,
    pub description: String
}

#[derive(Debug, Default, PartialEq, Clone)]
pub struct SliceSetSchema {
    pub id: i32,
    pub set_id: Uuid,
    pub timestamp_begin: DateTime<Utc>,
    pub timestamp_end: DateTime<Utc>,
    pub name: String,
    pub description: String
}

File: grpc_api/client_python/src/bbthings_grpc/resource/_schema.py
from typing import List, Union
from dataclasses import dataclass
from datetime import datetime
from uuid import UUID
from ..common.type_value import DataType, unpack_data, unpack_data_array


@dataclass
class ProcedureAcces:
    procedure: str
    roles: List[str]

    def from_response(r):
        return ProcedureAcces(r.procedure, r.roles)


@dataclass
class RoleAcces:
    role: str
    procedures: List[str]

    def from_response(r):
        return RoleAcces(r.role, r.procedures)


@dataclass
class ModelConfigSchema:
    id: int
    model_id: UUID
    index: int
    name: str
    value: Union[int, float, str, bool, None]
    category: str

    def from_response(r):
        value = unpack_data(r.config_bytes, DataType(r.config_type))
        return ModelConfigSchema(r.id, UUID(bytes=r.model_id), r.index, r.name, value, r.category)


@dataclass
class ModelSchema:
    id: UUID
    category: str
    name: str
    description: str
    data_type: List[DataType]
    configs: List[List[ModelConfigSchema]]

    def from_response(r):
        types = []
        for ty in r.data_type:
            types.append(DataType(ty))
        configs = []
        for conf_vec in r.configs:
            confs = []
            for conf in conf_vec.configs: confs.append(ModelConfigSchema.from_response(conf))
            configs.append(confs)
        return ModelSchema(UUID(bytes=r.id), r.category, r.name, r.description, types, configs)


@dataclass
class TagSchema:
    model_id: UUID
    tag: int
    name: str
    members: List[int]

    def from_response(r):
        return TagSchema(UUID(bytes=r.model_id), r.tag, r.name, r.members)


@dataclass
class TypeConfigSchema:
    id: int
    type_id: UUID
    name: str
    value_type: DataType
    category: str

    def from_response(r):
        return TypeConfigSchema(r.id, UUID(bytes=r.type_id), r.name, DataType(r.config_type), r.category)


@dataclass
class TypeSchema:
    id: UUID
    name: str
    description: str
    model_ids: List[UUID]
    configs: List[TypeConfigSchema]

    def from_response(r):
        models = []
        for model in r.model_ids: models.append(UUID(bytes=model))
        configs = []
        for conf in r.configs: configs.append(TypeConfigSchema.from_response(conf))
        return TypeSchema(UUID(bytes=r.id), r.name, r.description, models)


@dataclass
class DeviceConfigSchema:
    id: int
    device_id: UUID
    name: str
    value: Union[int, float, str, bool, None]
    category: str

    def from_response(r):
        value = unpack_data(r.config_bytes, DataType(r.config_type))
        return DeviceConfigSchema(r.id, UUID(bytes=r.device_id), r.name, value, r.category)


@dataclass
class GatewayConfigSchema:
    id: int
    gateway_id: UUID
    name: str
    value: Union[int, float, str, bool, None]
    category: str

    def from_response(r):
        value = unpack_data(r.config_bytes, DataType(r.config_type))
        return GatewayConfigSchema(r.id, UUID(bytes=r.gateway_id), r.name, value, r.category)


@dataclass
class DeviceSchema:
    id: UUID
    gateway_id: UUID
    serial_number: str
    name: str
    description: str
    type_id: UUID
    type_name: UUID
    model_ids: List[UUID]
    configs: List[DeviceConfigSchema]

    def from_response(r):
        configs = []
        for conf in r.configs: configs.append(DeviceConfigSchema.from_response(conf))
        model_ids = []
        for id in r.model_ids: model_ids.append(UUID(bytes=id))
        return DeviceSchema(UUID(bytes=r.id), UUID(bytes=r.gateway_id), r.serial_number, r.name, r.description, r.type_id, r.type_name, model_ids, configs)


@dataclass
class GatewaySchema:
    id: UUID
    serial_number: str
    name: str
    description: str
    type_id: UUID
    type_name: UUID
    model_ids: List[UUID]
    configs: List[GatewayConfigSchema]

    def from_response(r):
        configs = []
        for conf in r.configs: configs.append(GatewayConfigSchema.from_response(conf))
        model_ids = []
        for id in r.model_ids: model_ids.append(UUID(bytes=id))
        return GatewaySchema(UUID(bytes=r.id), r.serial_number, r.name, r.description, r.type_id, r.type_name, model_ids, configs)


@dataclass
class GroupModelSchema:
    id: UUID
    name: str
    category: str
    description: str
    model_ids: List[UUID]

    def from_response(r):
        models = []
        for model in r.model_ids: models.append(UUID(bytes=model))
        return GroupModelSchema(UUID(bytes=r.id), r.name, r.category, r.description, models)


@dataclass
class GroupDeviceSchema:
    id: UUID
    name: str
    category: str
    description: str
    device_ids: List[UUID]

    def from_response(r):
        devices = []
        for device in r.device_ids: devices.append(UUID(bytes=device))
        return GroupDeviceSchema(UUID(bytes=r.id), r.name, r.category, r.description, devices)


@dataclass
class GroupGatewaySchema:
    id: UUID
    name: str
    category: str
    description: str
    gateway_ids: List[UUID]

    def from_response(r):
        gateways = []
        for gateway in r.device_ids: gateways.append(UUID(bytes=gateway))
        return GroupGatewaySchema(UUID(bytes=r.id), r.name, r.category, r.description, gateways)


@dataclass
class SetMember:
    device_id: UUID
    model_id: UUID
    data_index: List[int]

    def from_response(r):
        return SetMember(UUID(bytes=r.device_id), UUID(bytes=r.model_id), list(r.data_index))


@dataclass
class SetSchema:
    id: UUID
    template_id: UUID
    name: str
    description: str
    members: List[SetMember]

    def from_response(r):
        members = []
        for member in r.members: members.append(SetMember.from_response(member))
        return SetSchema(UUID(bytes=r.id), UUID(bytes=r.template_id), r.name, r.description, members)


@dataclass
class SetTemplateMember:
    type_id: UUID
    model_id: UUID
    data_index: List[int]

    def from_response(r):
        return SetTemplateMember(UUID(bytes=r.type_id), UUID(bytes=r.model_id), list(r.data_index))


@dataclass
class SetTemplateSchema:
    id: UUID
    name: str
    description: str
    members: List[SetTemplateMember]

    def from_response(r):
        members = []
        for member in r.members: members.append(SetTemplateMember.from_response(member))
        return SetTemplateSchema(UUID(bytes=r.id), r.name, r.description, members)


@dataclass
class DataSchema:
    device_id: UUID
    model_id: UUID
    timestamp: datetime
    data: List[Union[int, float, str, bool, None]]
    tag: int

    def from_response(r):
        timestamp = datetime.fromtimestamp(r.timestamp/1000000.0)
        types = []
        for ty in r.data_type: types.append(DataType(ty))
        data = unpack_data_array(r.data_bytes, types)
        return DataSchema(UUID(bytes=r.device_id), UUID(bytes=r.model_id), timestamp, data, r.tag)


@dataclass
class DataSetSchema:
    set_id: UUID
    timestamp: datetime
    data: List[Union[int, float, str, bool, None]]
    tag: int

    def from_response(r):
        timestamp = datetime.fromtimestamp(r.timestamp/1000000.0)
        types = []
        for ty in r.data_type: types.append(DataType(ty))
        data = unpack_data_array(r.data_bytes, types)
        return DataSetSchema(UUID(bytes=r.set_id), timestamp, data, r.tag)


@dataclass
class BufferSchema:
    id: int
    device_id: UUID
    model_id: UUID
    timestamp: datetime
    data: List[Union[int, float, str, bool, None]]
    tag: int

    def from_response(r):
        timestamp = datetime.fromtimestamp(r.timestamp/1000000.0)
        types = []
        for ty in r.data_type: types.append(DataType(ty))
        data = unpack_data_array(r.data_bytes, types)
        return BufferSchema(r.id, UUID(bytes=r.device_id), UUID(bytes=r.model_id), timestamp, data, r.tag)


@dataclass
class BufferSetSchema:
    ids: List[int]
    set_id: UUID
    timestamp: datetime
    data: List[Union[int, float, str, bool, None]]
    tag: int

    def from_response(r):
        timestamp = datetime.fromtimestamp(r.timestamp/1000000.0)
        types = []
        for ty in r.data_type: types.append(DataType(ty))
        data = unpack_data_array(r.data_bytes, types)
        return BufferSchema(r.ids, UUID(bytes=r.set_id), timestamp, data, r.tag)


@dataclass
class SliceSchema:
    id: int
    device_id: UUID
    model_id: UUID
    timestamp_begin: datetime
    timestamp_end: datetime
    name: str
    description: str

    def from_response(r):
        timestamp_begin = datetime.fromtimestamp(r.timestamp_begin/1000000.0)
        timestamp_end = datetime.fromtimestamp(r.timestamp_end/1000000.0)
        return SliceSchema(r.id, UUID(bytes=r.device_id), UUID(bytes=r.model_id), timestamp_begin, timestamp_end, r.name, r.description)


@dataclass
class SliceSetSchema:
    id: int
    set_id: UUID
    timestamp_begin: datetime
    timestamp_end: datetime
    name: str
    description: str

    def from_response(r):
        timestamp_begin = datetime.fromtimestamp(r.timestamp_begin/1000000.0)
        timestamp_end = datetime.fromtimestamp(r.timestamp_end/1000000.0)
        return SliceSetSchema(r.id, UUID(bytes=r.set_id), timestamp_begin, timestamp_end, r.name, r.description)
